lib/resque/durable/background_heartbeat.rb

```diff
@@ -0,0 +1,83 @@
+require 'thread'
+
+module Resque
+  module Durable
+
+    # Creates a background thread to regularly heartbeat the queue audit.
+    class BackgroundHeartbeat
+      DEFAULT_INTERVAL = 15
+
+      def initialize(queue_audit, interval = nil)
+        @queue_audit  = queue_audit
+        @last_timeout = nil
+        @interval     = interval || DEFAULT_INTERVAL
+        @mutex        = Mutex.new
+        @cv           = ConditionVariable.new
+        @stop         = false
+        @thread       = nil
+      end
+
+      class << self
+        # only a separate method for easy stubbing
+        def exit_now!
+          abort
+        end
+      end
+
+      def with_heartbeat
+        start!
+        yield
+      ensure
+        stop_and_wait!
+      end
+
+      def heartbeat!
+        @last_timeout ||= @queue_audit.timeout_at
+        @last_timeout = @queue_audit.optimistic_heartbeat!(@last_timeout)
```

Comment:
I might be missing something. `QueueAudit#timeout_at` isn't set by `optimistic_heartbeat` since it's an update_all, `QueueAudit#heartbeat!` isn't called, and it doesn't seem to default to anything. Would the statement above return nil the first time (which is ok, just unnecessary based on the default)?